File: js-src/src/storage/domain-store/BaseStore.ts
import { Change } from "../../model/Change";
import { Entity, EntityData, SerializedEntityData, dumpToDict, loadFromDict, transformedEntity } from "../../model/Entity";
import { Delta } from "../../model/Delta";

// a searchFilter is a dict, where there can be an id, a parent_id, a type or any other property
export interface SearchFilter {
    id?: string;
    parentId?: string;
    [key: string]: any;
}

export interface SerializedStoreData {
    entities: SerializedEntityData[];
}

export class IrrationalStorageOperation extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'IrrationalStorageOperation';
    }
}

export abstract class Store {
    onChanges: ((delta: Delta, origin?: string) => void) | null = null;
    protected _applyingInternally: boolean = false;

    abstract insertOne(entity: Entity<EntityData>): Change;

    abstract find(filter: SearchFilter): Generator<Entity<EntityData>>;

    abstract findOne(filter: SearchFilter): Entity<EntityData> | undefined;

    abstract updateOne(entity: Entity<EntityData>): Change;

    abstract removeOne(entity: Entity<EntityData>): Change;

    // Batch operations (inefficient implementations)
    insert(entities: Entity<EntityData>[]): Change[] {
        return entities.map(entity => this.insertOne(entity))
    }
    remove(entities: Entity<EntityData>[]): Change[] {
        return entities.map(entity => this.removeOne(entity))
    }
    update(entities: Entity<EntityData>[]): Change[] {
        return entities.map(entity => this.updateOne(entity))
    }
    /** Apply a single change, firing onChanges with the given origin. */
    applyChange(change: Change, origin?: string): Change {
        this._applyingInternally = true;
        let appliedChange: Change;
        try {
            appliedChange = this._applyChangeCore(change);
        } finally {
            this._applyingInternally = false;
        }
        if (this.onChanges && !appliedChange.isEmpty()) {
            this.onChanges(Delta.fromChanges([appliedChange]), origin);
        }
        return appliedChange;
    }

    private _applyChangeCore(change: Change): Change {
        if (change.isCreate()) {
            let entityState = loadFromDict(change.forwardComponent as SerializedEntityData);
            return this.insertOne(entityState);
        } else if (change.isUpdate()) {
            let entity = this.findOne({ id: change.entityId });
            if (entity === undefined) {
                throw new IrrationalStorageOperation(`Last state retreival error for ${change.data} (type update)`);
            }
            entity = transformedEntity(entity, change);
            return this.updateOne(entity);
        } else if (change.isDelete()) {
            let entity = this.findOne({ id: change.entityId });
            if (entity === undefined) {
                throw new IrrationalStorageOperation(`Last state retreival error for ${change.data} (type delete)`);
            }
            return this.removeOne(entity);
        }
        return change;
    }

    data(): SerializedStoreData {
        const entities = Array.from(this.find({}));
        return { entities: entities.map(entity => dumpToDict(entity)) };
    }

    abstract clear(): void;

    protected _loaded: boolean = false;

    /** Load initial data into an empty (or cleared) store. Can only be called once, or again after clear(). */
    loadData(entities: Entity<EntityData>[], origin?: string): Delta {
        if (this._loaded) {
            throw new Error('Store already has data loaded. Call clear() before loading again.');
        }
        this._applyingInternally = true;
        const changes: Change[] = [];
        try {
            for (const entity of entities) {
                changes.push(this.insertOne(entity));
            }
        } finally {
            this._applyingInternally = false;
            this._loaded = true;
        }

        const delta = Delta.fromChanges(changes);
        if (this.onChanges && !delta.isEmpty()) {
            this.onChanges(delta, origin);
        }
        return delta;
    }

    /** Apply a delta, firing onChanges once at the end with the given origin. */
    applyDelta(delta: Delta, origin?: string, skipIrrationalOperations: boolean = false): Delta {
        this._applyingInternally = true;

        const appliedDelta = new Delta({});
        try {
            for (const change of delta.changes()) {
                try {
                    const appliedChange = this._applyChangeCore(change);
                    appliedDelta.addChangeFromData(appliedChange.data);
                } catch (error) {
                    if (!skipIrrationalOperations || !(error instanceof IrrationalStorageOperation)) {
                        throw error;
                    }
                }
            }
        } finally {
            this._applyingInternally = false;
        }

        if (this.onChanges && !appliedDelta.isEmpty()) {
            this.onChanges(appliedDelta, origin);
        }
        return appliedDelta;
    }
}

File: py-src/fusion/storage/base_store.py
from collections.abc import Callable, Generator
from typing import Any

from fusion import Entity
from fusion.libs.model import load_from_dict, transformed_entity
from fusion.storage.change import Change
from fusion.storage.delta import Delta


class Store:
    _applying_internally: bool = False

    def __init__(self) -> None:
        self._on_changes_callbacks: list[Callable[[Delta, str | None], None]] = []

    # --- on_changes multi-handler API ---

    def add_on_changes_callback(
        self, callback: Callable[[Delta, str | None], None]
    ) -> None:
        self._on_changes_callbacks.append(callback)

    def remove_on_changes_callback(
        self, callback: Callable[[Delta, str | None], None]
    ) -> None:
        self._on_changes_callbacks.remove(callback)

    def _fire_on_changes(self, delta: Delta, origin: str | None = None) -> None:
        for cb in self._on_changes_callbacks:
            cb(delta, origin)

    @property
    def on_changes(self) -> Callable[[Delta, str | None], None] | None:
        """Backward-compat: returns the fire method if callbacks are registered."""
        if self._on_changes_callbacks:
            return self._fire_on_changes
        return None

    @on_changes.setter
    def on_changes(self, callback: Callable[[Delta, str | None], None] | None) -> None:
        """Backward-compat: replaces all callbacks with a single one."""
        self._on_changes_callbacks.clear()
        if callback is not None:
            self._on_changes_callbacks.append(callback)

    def insert_one(self, entity: Entity) -> Change:
        raise NotImplementedError

    def find(self, **filter: Any) -> Generator[Any, None, None]:
        raise NotImplementedError

    def find_one(self, **filter: Any) -> Any | None:
        found = self.find(**filter)

        for f in found:
            return f

        return None

    def update_one(self, entity: Entity) -> Change:
        raise NotImplementedError

    def remove_one(self, entity: Entity) -> Change:
        raise NotImplementedError

    # Batch operations (inefficient implementations)
    def insert(self, batch: list[Entity]) -> list[Change]:
        return [self.insert_one(entity) for entity in batch]

    def remove(self, batch: list[Entity]) -> list[Change]:
        return [self.remove_one(entity) for entity in batch]

    def update(self, batch: list[Entity]) -> list[Change]:
        return [self.update_one(entity) for entity in batch]

    def apply_change(self, change: Change, origin: str | None = None) -> Change:
        self._applying_internally = True
        try:
            applied = self._apply_change_core(change)
        finally:
            self._applying_internally = False

        if self._on_changes_callbacks and not applied.is_empty():
            self._fire_on_changes(Delta.from_changes([applied]), origin)
        return applied

    def clear(self) -> None:
        raise NotImplementedError

    _loaded: bool = False

    def load_data(self, entities: list[Entity], origin: str | None = None) -> Delta:
        """Load initial data into an empty (or cleared) store. Can only be called once, or again after clear()."""
        if self._loaded:
            raise RuntimeError(
                "Store already has data loaded. Call clear() before loading again."
            )
        self._applying_internally = True
        changes: list[Change] = []
        try:
            for entity in entities:
                changes.append(self.insert_one(entity))
        finally:
            self._applying_internally = False
            self._loaded = True

        return Delta.from_changes(changes)

    def apply_delta(self, delta: Delta, origin: str | None = None) -> Delta:
        """Apply all changes in a Delta to the store. Returns the applied delta."""
        self._applying_internally = True

        applied = Delta()
        try:
            for change in delta.changes():
                applied_change = self._apply_change_core(change)
                applied.add_change(applied_change)
        finally:
            self._applying_internally = False

        if self._on_changes_callbacks and not applied.is_empty():
            self._fire_on_changes(applied, origin)
        return applied

    def _apply_change_core(self, change: Change) -> Change:
        """Apply a single change without firing on_changes."""
        if change.is_create():
            entity = load_from_dict(dict(change.forward_component))
            return self.insert_one(entity)
        elif change.is_update():
            entity = self.find_one(id=change.entity_id)
            if entity is None:
                raise ValueError(f"Entity {change.entity_id} not found for update")
            entity = transformed_entity(entity, change)
            return self.update_one(entity)
        elif change.is_delete():
            entity = self.find_one(id=change.entity_id)
            if entity is None:
                raise ValueError(f"Entity {change.entity_id} not found for delete")
            return self.remove_one(entity)
        return change
